File: 二叉树/输出二叉树.java
package 二叉树;

import java.util.ArrayList;
import java.util.List;

public class 输出二叉树 {
    static int height;
    List<List<String>> res = new ArrayList<>();


    public List<List<String>> printTree(TreeNode root) {

        height = depth(root) - 1;
        int m = height + 1;
        int n = (int) Math.pow(2, height + 1) - 1;
        //List你只能一个一个加元素了
        for (int i = 0; i < m; i++) {
            List<String> row = new ArrayList<>();
            for (int j = 0; j < n; j++) {
                row.add("");
            }
            res.add(row);
        }
        dfs(root, 0, (n - 1) / 2);
        return res;
    }

    public int depth(TreeNode root) {
        if (root == null) return 0;
        return Math.max(depth(root.left), depth(root.right)) + 1;
    }

    public void dfs(TreeNode root, int i, int j) {
        if (root == null) return;
        //list修改元素用set
        res.get(i).set(j, Integer.toString(root.val));
        int element = (int) Math.pow(2, height - i - 1);
        dfs(root.left, i + 1, j - element);
        dfs(root.right, i + 1, j + element);
    }
}

File: 二叉树/输出二叉树.go
package 二叉树

import (
	"math"
	"strconv"
)

//这也算dfs哦
func printTree(root *TreeNode) [][]string {
	height := depth(root) - 1
	m := height + 1
	n := int(math.Pow(2.0, float64(height+1))) - 1
	res := make([][]string, m)
	for i := 0; i < m; i++ {
		res[i] = make([]string, n)
	}
	var dfs func(*TreeNode, int, int)
	dfs = func(root *TreeNode, curRow, curCol int) {
		if root == nil {
			return
		}
		res[curRow][curCol] = strconv.Itoa(root.Val)
		element := int(math.Pow(2.0, float64(height-curRow-1)))
		dfs(root.Left, curRow+1, curCol-element)
		dfs(root.Right, curRow+1, curCol+element)
	}
	dfs(root, 0, (n-1)/2)
	return res
}

func depth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	l := depth(root.Left)
	r := depth(root.Right)
	return max(l, r) + 1
}
